plasTeX/ConfigManager.py

```diff
@@ -0,0 +1,181 @@
+from typing import Dict, List, Generic, TypeVar, Any, Optional, Type, KeysView, Union
+from configparser import ConfigParser
+import shlex
+
+T = TypeVar('T')
+
+def formatDefault(s: Any):
+    return " [" + str(s).replace("%", "%%") + "]"
+
+class Option(Generic[T]):
+    def __init__(self, description: str, options: str, default: T):
+        self.description = description
+        self.options = options.split(" ")
+        self.name = self.options[0].lstrip("-")
+        self._value = default
+
+    def registerArgparse(self, group: ArgumentGroup):
+        group.add_argument(*self.options, dest=self.name, type=self.valueType(), help=self.description + formatDefault(self.value))
+
+    def valueType(self) -> Type:
+        return type(self.value)
+
+    def updateFromDict(self, data: Dict["str", Any]):
+        value = data.get(self.name)
+        if value is not None:
+            self.value = value
+
+    def setFromString(self, string: str):
+        self.value = self.valueType()(string)
+
+    @property
+    def value(self) -> T:
+        return self._value
+
+    @value.setter
+    def value(self, value: T):
+        if type(value) != self.valueType():
+            raise TypeError("Encountered {} when expecting {}".format(type(value).__name__, self.valueType().__name__))
+
+        self._value = value
+
+class BooleanOption(Option[bool]):
+    def registerArgparse(self, group: ArgumentGroup):
+        enables = [x for x in self.options if x[0] != "!"]
+        disables = [x[1:] for x in self.options if x[0] == "!"]
+
+        group.add_argument(*enables, dest=self.name, help=self.description + formatDefault(self.value), action='store_true', default=None)
+        if len(disables) > 0:
+            group.add_argument(*disables, dest=self.name, help=self.description + formatDefault(self.value), action='store_false', default=None)
+
+class MultiStringOption(Option[List[str]]):
+    def registerArgparse(self, group: ArgumentGroup):
+        group.add_argument(*self.options, dest=self.name, type=str, nargs="*", help=self.description, action="append")
+
+    def setFromString(self, string: str):
+        self.value.extend(shlex.split(string))
+
+    def updateFromDict(self, data: Dict["str", Any]):
+        value = data.get(self.name)
+        if value is not None:
+            for entry in value:
+                self._value.extend(entry)
+
+class DictOption(ABC, Option[Dict[str, T]]):
+    @classmethod
+    @abstractmethod
+    def entryFromString(cls, entry: str) -> T:
+        """This convers a string into T"""
+
+    def set(self, key: str, value: str):
+        self.value[key] = self.entryFromString(value)
+
+    def setFromString(self, string: str):
+        raise ValueError("Calling setFromString on a DictOption. The individual items should be inserted with set.")
+
+    def updateFromDict(self, data: Dict["str", Any]):
+        entries = data.get(self.name)
+        if entries is not None:
+            for key, val in entries:
```

Comment:
What is the type of `entries` here? Is this function covered by a test? Again it would help to give a default value on line 79 which would get rid of the test and create an empty loop run.

Reply:
It's a dict. This is not covered by a test, since DictOption should not be directly instantiated. I should add tests for Links and Counters, the latter of which uses the default `updateFromDict` and the former uses a custom one.

Reply:
This is what I thought, and I don't see how this code could work, hence the need to test. I would understand `for key, val in entries.items():`.

Reply:
You are absolutely right.